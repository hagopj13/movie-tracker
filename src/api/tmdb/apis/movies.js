// @flow
import axios from '../axiosInstance';

export type GetDiscoverMoviesParams = {
  page?: number,
  sort_by: string,
  'vote_count.gte'?: number,
  with_genres?: string,
  'primary_release_date.gte'?: string,
  'primary_release_date.lte'?: string,
};
export const getDiscoverMovies = (params: GetDiscoverMoviesParams) =>
  axios.get('/discover/movie', { params });

export type GetUpcomingMoviesParams = {
  page?: number,
};
export const getUpcomingMovies = (params: GetUpcomingMoviesParams) =>
  axios.get('/movie/upcoming', { params });

export type SearchMoviesParams = {
  page?: number,
  query: string,
};
export const searchMovies = (params: SearchMoviesParams) => axios.get('/search/movie', { params });

export const getMovieDetails = (id: string, sessionId?: string) =>
  axios.get(`/movie/${id}`, {
    params: {
      session_id: sessionId,
      append_to_response: 'credits,keywords,reviews,recommendations,account_states',
    },
  });

export const getMovieUserState = (id: string, sessionId: string) =>
  axios.get(`/movie/${id}/account_states?session_id=${sessionId}`);

export const setIsMovieFavorite = (
  id: string,
  isFavorite: boolean,
  sessionId: string,
  accountId: string,
) =>
  axios.post(`/account/${accountId}/favorite?session_id=${sessionId}`, {
    media_type: 'movie',
    media_id: id,
    favorite: isFavorite,
  });

export const setIsMovieInWatchlist = (
  id: string,
  isInWatchlist: boolean,
  sessionId: string,
  accountId: string,
) =>
  axios.post(`/account/${accountId}/watchlist?session_id=${sessionId}`, {
    media_type: 'movie',
    media_id: id,
    watchlist: isInWatchlist,
  });

export const rateMovie = (id: string, rating: number, sessionId: string) =>
  axios.post(`/movie/${id}/rating?session_id=${sessionId}`, { value: rating });

export const deleteMovieRating = (id: string, sessionId: string) =>
  axios.delete(`/movie/${id}/rating?session_id=${sessionId}`);
