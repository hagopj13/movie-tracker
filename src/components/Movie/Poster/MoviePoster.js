// @flow
import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { makeStyles } from '@material-ui/core/styles';
import CardMedia from '@material-ui/core/CardMedia';
import ImageIcon from '@material-ui/icons/Image';

import configSelectors from 'store/config/config.selectors';

type Props = {
  posterPath?: string,
  fullPosterPath: string,
  height?: number,
  width?: number,
};

const useStyles = makeStyles((theme) => ({
  brokenMediaContainer: {
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: theme.palette.grey[200],
  },
}));

export const MoviePoster = (props: Props) => {
  const { posterPath, fullPosterPath, height, width } = props;

  const classes = useStyles();

  const isImageFound = posterPath && fullPosterPath;
  const imageStyle = {};
  if (height) {
    imageStyle.height = height;
  }
  if (width) {
    imageStyle.width = width;
  }

  return isImageFound ? (
    <CardMedia image={fullPosterPath} style={imageStyle} />
  ) : (
    <div className={classes.brokenMediaContainer} style={imageStyle}>
      <ImageIcon fontSize="large" />
    </div>
  );
};

MoviePoster.defaultProps = {
  posterPath: null,
  height: 0,
  width: 0,
};

const mapStateToProps = createStructuredSelector({
  fullPosterPath: (state, ownProps: Props) =>
    configSelectors.createPosterFullPathSelector(ownProps.posterPath)(state),
});

export default connect(mapStateToProps)(MoviePoster);
