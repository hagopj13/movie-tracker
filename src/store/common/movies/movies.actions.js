// @flow
import typeGenerator from 'store/common/utils/typeGenerator';
import actionCreatorGenerator from 'store/common/utils/actionCreatorGenerator';
import type { Movie, Results } from 'types';

import MoviesActionTypes from './movies.types';

const fetchMovies = (namespace) => () => ({
  type: typeGenerator(namespace, MoviesActionTypes.FETCH_MOVIES),
});

const fetchMoviesStart = (namespace) => () => ({
  type: typeGenerator(namespace, MoviesActionTypes.FETCH_MOVIES_START),
});

const fetchMoviesSuccess = (namespace) => (results: Results<Movie>) => ({
  type: typeGenerator(namespace, MoviesActionTypes.FETCH_MOVIES_SUCCESS),
  payload: { results },
});

const fetchMoviesFailure = (namespace) => (error: string) => ({
  type: typeGenerator(namespace, MoviesActionTypes.FETCH_MOVIES_FAILURE),
  payload: { error },
});

const fetchMoreMovies = (namespace) => () => ({
  type: typeGenerator(namespace, MoviesActionTypes.FETCH_MORE_MOVIES),
});

const fetchMoreMoviesStart = (namespace) => () => ({
  type: typeGenerator(namespace, MoviesActionTypes.FETCH_MORE_MOVIES_START),
});

const fetchMoreMoviesSuccess = (namespace) => (results: Results<Movie>) => ({
  type: typeGenerator(namespace, MoviesActionTypes.FETCH_MORE_MOVIES_SUCCESS),
  payload: { results },
});

const fetchMoreMoviesFailure = (namespace) => (error: string) => ({
  type: typeGenerator(namespace, MoviesActionTypes.FETCH_MORE_MOVIES_FAILURE),
  payload: { error },
});

type Action = {
  type: string,
  payload?: any,
};

type MoviesActionCreators = {
  fetchMovies: () => Action,
  fetchMoviesStart: () => Action,
  fetchMoviesSuccess: (results: Results<Movie>) => Action,
  fetchMoviesFailure: (error: string) => Action,
  fetchMoreMovies: () => Action,
  fetchMoreMoviesStart: () => Action,
  fetchMoreMoviesSuccess: (results: Results<Movie>) => Action,
  fetchMoreMoviesFailure: (error: string) => Action,
};

export default (namespace: string): MoviesActionCreators =>
  actionCreatorGenerator<MoviesActionCreators>(namespace, {
    fetchMovies,
    fetchMoviesStart,
    fetchMoviesSuccess,
    fetchMoviesFailure,
    fetchMoreMovies,
    fetchMoreMoviesStart,
    fetchMoreMoviesSuccess,
    fetchMoreMoviesFailure,
  });
