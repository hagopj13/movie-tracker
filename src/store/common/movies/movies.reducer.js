// @flow
import reducerGenerator from 'store/common/utils/reducerGenerator';
import type { Movie } from 'types';

import MoviesActionTypes from './movies.types';

type State = {
  list: Movie[],
  pagination: {
    page: number,
    totalPages: number,
    totalResults: number,
  } | null,
};

export const defaultInitialState: State = {
  list: [],
  pagination: null,
};

const fetchMoviesStart = () => defaultInitialState;

const fetchMoviesSuccess = (state: State, action) => ({
  ...state,
  list: action.payload.results.list,
  pagination: action.payload.results.pagination,
});

const fetchMoreMoviesSuccess = (state: State, action) => ({
  ...state,
  list: state.list.concat(action.payload.results.list),
  pagination: {
    ...state.pagination,
    page: action.payload.results.pagination.page,
  },
});

const moviesActionHandler = {
  [MoviesActionTypes.FETCH_MOVIES_START]: fetchMoviesStart,
  [MoviesActionTypes.FETCH_MOVIES_SUCCESS]: fetchMoviesSuccess,
  [MoviesActionTypes.FETCH_MORE_MOVIES_SUCCESS]: fetchMoreMoviesSuccess,
};

export default (namespace: string, initialState?: State) =>
  reducerGenerator(namespace, moviesActionHandler, initialState ?? defaultInitialState);
