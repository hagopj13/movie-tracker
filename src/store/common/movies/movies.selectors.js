import { createSelector } from 'reselect';

const selectList = (state) => state.list;

const selectPagination = (state) => state.pagination;

const selectCurrentPage = createSelector([selectPagination], (pagination) => pagination?.page ?? 1);

const selectTotalPages = createSelector(
  [selectPagination],
  (pagination) => pagination?.totalPages ?? 0,
);

const selectTotalResults = createSelector(
  [selectPagination],
  (pagination) => pagination?.totalResults ?? 0,
);

export default {
  selectList,
  selectCurrentPage,
  selectTotalPages,
  selectTotalResults,
};
