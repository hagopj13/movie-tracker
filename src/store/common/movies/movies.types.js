export default {
  FETCH_MOVIES: 'FETCH_MOVIES',
  FETCH_MOVIES_START: 'FETCH_MOVIES_START',
  FETCH_MOVIES_SUCCESS: 'FETCH_MOVIES_SUCCESS',
  FETCH_MOVIES_FAILURE: 'FETCH_MOVIES_FAILURE',
  FETCH_MORE_MOVIES: 'FETCH_MORE_MOVIES',
  FETCH_MORE_MOVIES_START: 'FETCH_MORE_MOVIES_START',
  FETCH_MORE_MOVIES_SUCCESS: 'FETCH_MORE_MOVIES_SUCCESS',
  FETCH_MORE_MOVIES_FAILURE: 'FETCH_MORE_MOVIES_FAILURE',
};
