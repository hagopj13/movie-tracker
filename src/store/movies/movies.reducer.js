import { combineReducers } from 'redux';

import discoverReducer from './discover/discover.reducer';

export default combineReducers({
  discover: discoverReducer,
});
