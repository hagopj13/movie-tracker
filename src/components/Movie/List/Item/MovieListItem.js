// @flow
import React from 'react';
import { useHistory } from 'react-router-dom';
import moment from 'moment';
import { makeStyles } from '@material-ui/core/styles';
import Card from '@material-ui/core/Card';
import CardActionArea from '@material-ui/core/CardActionArea';
import CardContent from '@material-ui/core/CardContent';
import Typography from '@material-ui/core/Typography';
import Rating from '@material-ui/lab/Rating';

import MoviePoster from 'components/Movie/Poster/MoviePoster';
import type { Movie } from 'types';

type Props = {
  movie: Movie,
};

const useStyles = makeStyles((theme) => ({
  root: {
    width: 200,
    minWidth: 200,
    cursor: 'pointer',
  },
  title: {
    fontSize: 16,
  },
  rating: {
    display: 'flex',
    marginTop: theme.spacing(0.75),
  },
  ratingText: {
    marginLeft: theme.spacing(0.5),
  },
}));

const MovieListItem = (props: Props) => {
  const { movie } = props;

  const classes = useStyles();
  const history = useHistory();

  const formattedDate = movie.releaseDate
    ? moment(movie.releaseDate).format('MMM D, YYYY')
    : 'Unknown release date';

  const handleItemClick = () => {
    history.push(`/movies/${movie.id}`);
  };

  const renderRating = () => {
    return movie.voteCount ? (
      <>
        <Rating value={movie.voteAverage} precision={0.5} readOnly />
        <Typography className={classes.ratingText}>{movie.voteAverage}</Typography>
      </>
    ) : (
      <Typography>Not rated yet</Typography>
    );
  };

  return (
    <Card className={classes.root} elevation={3} onClick={handleItemClick}>
      <CardActionArea>
        <MoviePoster posterPath={movie.posterPath} height={300} width={200} />
        <CardContent>
          <Typography className={classes.title} gutterBottom variant="h6">
            {movie.title}
          </Typography>
          <Typography>{formattedDate}</Typography>
          <div className={classes.rating}>{renderRating()}</div>
        </CardContent>
      </CardActionArea>
    </Card>
  );
};

export default MovieListItem;
