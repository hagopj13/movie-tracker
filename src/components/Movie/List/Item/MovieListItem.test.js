import React from 'react';
import { createShallow } from '@material-ui/core/test-utils';
import Card from '@material-ui/core/Card';
import Typography from '@material-ui/core/Typography';
import Rating from '@material-ui/lab/Rating';

import movies from 'store/fixtures/movies';
import MoviePoster from 'components/Movie/Poster/MoviePoster';

import MovieListItem from './MovieListItem';

const mockHistoryPush = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}));

describe('MovieListItem component', () => {
  let shallow;
  let movie;
  let wrapper;

  beforeAll(() => {
    shallow = createShallow();
  });

  beforeEach(() => {
    jest.clearAllMocks();
    [movie] = movies.list;
    const mockProps = {
      movie,
    };
    wrapper = shallow(<MovieListItem {...mockProps} />);
  });

  it('should render the MovieListItem component correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render the movie title', () => {
    expect(wrapper.find(Typography).at(0).text()).toBe(movie.title);
  });

  it('should pass the posterPath to MoviePoster', () => {
    expect(wrapper.find(MoviePoster).prop('posterPath')).toBe(movie.posterPath);
  });

  it('should render Rating if voteCount is greater than 0', () => {
    expect(wrapper.exists(Rating)).toBe(true);
  });

  it('should push the correct url to history when movie is clicked', () => {
    wrapper.find(Card).simulate('click');
    expect(mockHistoryPush).toHaveBeenCalledTimes(1);
    expect(mockHistoryPush).toHaveBeenLastCalledWith(`/movies/${movie.id}`);
  });
});
