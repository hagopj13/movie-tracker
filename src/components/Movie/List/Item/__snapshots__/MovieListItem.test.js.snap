// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieListItem component should render the MovieListItem component correctly 1`] = `
<WithStyles(ForwardRef(Card))
  className="makeStyles-root-1"
  elevation={3}
  onClick={[Function]}
>
  <WithStyles(ForwardRef(CardActionArea))>
    <Connect(MoviePoster)
      height={300}
      posterPath="/xBHvZcjRiWyobQ9kxBhO6B2dtRI.jpg"
      width={200}
    />
    <WithStyles(ForwardRef(CardContent))>
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-title-2"
        gutterBottom={true}
        variant="h6"
      >
        Ad Astra
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Typography))>
        Sep 17, 2019
      </WithStyles(ForwardRef(Typography))>
      <div
        className="makeStyles-rating-3"
      >
        <WithStyles(ForwardRef(Rating))
          precision={0.5}
          readOnly={true}
          value={3}
        />
        <WithStyles(ForwardRef(Typography))
          className="makeStyles-ratingText-4"
        >
          3
        </WithStyles(ForwardRef(Typography))>
      </div>
    </WithStyles(ForwardRef(CardContent))>
  </WithStyles(ForwardRef(CardActionArea))>
</WithStyles(ForwardRef(Card))>
`;
